cloudinit/url_helper.py

```diff
@@ -81,9 +81,12 @@ def combine_single(url, add_on):
     return url
 
 
-def read_file_or_url(url, timeout=5, retries=10,
-                     headers=None, data=None, sec_between=1, ssl_details=None,
-                     headers_cb=None, exception_cb=None):
+def read_file_or_url(url, *args, data=None, **kwargs):
```

Comment:
If we change the `if data` into `if kwargs.get("data"):` then we wouldn't have to special-case it.

Reply:
Doesn't the old signature of `read_file_or_url()` and of `readurl()` allow them to be called with data as a non-keyword argument? This change would fail to log a warning for an existing call that looks like this:
```
results = read_file_or_url("file://tmp/myfile", 5, 10, None, "This is my data")
```
By only checking kwargs for the presence of `data`, you miss the case where it was passed as a positional argument.

Reply:
Good thought @jasonzio thanks. The only place in cloudinit providing positional params other than url was in util:read_seeded. That old beastie had invalid positional params to read_file_or_url from around 2011 and hadn't yet surfaced as a bug. A closer look lead me to push two more commits on this branch to resolve that.